### DISCUSSION TITLE
Can gSig be an even Number ?

### DISCUSSION THREAD

**Discussion by bashsyr** [flatironinstitute/CaImAn#1237] [Q&A] [answered]
Hello,
I've read somewhere when I started with Caiman, that gSig is the size of a gaussian filter and that it should be odd (for math reasons).
But then Yesterday I was watching the Caiman Intro Video ([This ](https://www.youtube.com/watch?v=5APzPRbzUIA)one) and they used gSig = 4, and now I'm confused.
Can it be bee even??,  it it recommended??  
A little explanation would be Great
Thank you :)

**Comment by bashsyr** [comment]
how about gSig_filt ?
in mesmerize they say it has to be odd, but then I saw here in some discussions that even values are used

**Reply by kushalkolar** [reply]
mesmerize just gives the params directly to caiman, it does not put any constraints?

**Reply by bashsyr** [reply]
Yes, But in the Motioncorrection GUI, you can only choose odd values for gSig_filt, whcih confirmed my theory of them having to be odd
![image](https://github.com/flatironinstitute/CaImAn/assets/74191284/71522abe-19d3-4822-9e0d-c885700912a5)

**Reply by kushalkolar** [reply]
You can change the slider step size and see if opencv throws errors.

**Reply by kushalkolar** [reply]
In the caiman demo it says "size of high pass spatial filtering, used in 1p data", I assumed that means kernel size, and opencv doesn't accept even sized kernels.

**Reply by bashsyr** [reply]
Yeah, it worked, No errors, and now I have more granularity, I'll have to redo my testing now to see if even sized gSig_filt generally have any disadvantages when it comes to capturing motion but it seems to work, Thanks

**Reply by bashsyr** [reply]
So I did some testing, and it seems to work fine, Videos can look stable for even gSig_filt and there seem to be no weird behavior or anything.
![shifts_27](https://github.com/flatironinstitute/CaImAn/assets/74191284/99a5f62f-7d6b-4d9b-ba5a-cc873306efd1)
The found movement seems to be a bit more correlated for odd value alone (or even alone ) vs both together, but it seems to work fine
Here's correlation between all:
![Figure_27](https://github.com/flatironinstitute/CaImAn/assets/74191284/ec5ac85f-8e7f-4323-815f-b44c59801c1e)
Here are just the even values:
![Figure_27_even](https://github.com/flatironinstitute/CaImAn/assets/74191284/72122c31-636b-4c1f-a5eb-3f87d10d201f)

**Comment by EricThomson** [comment, marked as answer]
`gSig` and `gsig_filt` can be even, `gSiz` is odd.  But even if you don't make it odd, the code will force it to be odd on the back end.

All the `sig` params are gaussian filter sigma (standard deviation) parameters. gSiz is a bit all over the place in the code, but in the context of initialization it is the size in pixels of the gaussian kernel that you convolve the images with -- you basically want to be sure that the kernel is large enough to contain all the energy in the convolution kernel.  

We use opencv for gaussian blurring. The params are discussed here:
https://docs.opencv.org/4.x/d4/d86/group__imgproc__filter.html#gaabe8c836e97159a9193fb0b11ac52cf1

Their `ksize` is our `gSiz`. Their `sigmaX` and `sigmaY` are our `gSig` values.

Some places in the code `gSiz` is described as the width of a neuron, and this is just inaccurate. We need to better document all this.
